tests/code_update.py

```diff
@@ -0,0 +1,238 @@
+    result = infra.proc.ccall(
+        "./memberclient",
+        "vote",
+        "--accept",
+        "--cert=member1_cert.pem",
+        "--privk=member1_privk.pem",
+        f"--host={primary.host}",
+        f"--port={primary.tls_port}",
+        f"--id={proposal_id}",
+        "--ca=networkcert.pem",
+        "--sign",
+    )
+    j_result = json.loads(result.stdout)
+    assert not j_result["result"]
+
+    result = infra.proc.ccall(
+        "./memberclient",
+        "vote",
+        "--accept",
+        "--cert=member2_cert.pem",
+        "--privk=member2_privk.pem",
+        f"--host={primary.host}",
+        f"--port={primary.tls_port}",
+        f"--id={proposal_id}",
+        "--ca=networkcert.pem",
+        "--sign",
+    )
+    j_result = json.loads(result.stdout)
+    assert j_result["result"]
+
+
+def add_new_code(primary):
+    new_code_id = create_new_code_version(primary)
+    LOG.debug(f"New code id: {new_code_id}")
+
+    # first propose adding the new code id
+    result = infra.proc.ccall(
+        "./memberclient",
+        "add_code",
+        "--cert=member1_cert.pem",
+        "--privk=member1_privk.pem",
+        f"--host={primary.host}",
+        f"--port={primary.tls_port}",
+        "--ca=networkcert.pem",
+        f"--new_code_id={new_code_id}",
+    )
+
+    vote_to_accept(primary, 0)
+
+
+def create_node_using_new_code(network, args, node_id):
+    # add a node using unsupported code
+    assert network.create_and_add_node(
+        args.patched_file_name, args, node_id, False
+    ) == (False, infra.jsonrpc.ErrorCode.CODE_ID_NOT_FOUND)
+
+
+def run(args):
+    hosts = ["localhost", "localhost"]
+
+    with infra.ccf.network(
+        hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        primary, others = network.start_and_join(args)
+
+        forwarded_args = {
+            arg: getattr(args, arg) for arg in infra.ccf.Network.node_args_to_forward
+        }
+
+        res, new_node, new_node_id = network.create_and_add_node(
+            args.package, args, 2, True
+        )
+        new_node.join_network()
+
+        # try to add a node using unsupported code
+        assert network.create_and_add_node(args.patched_file_name, args, 3, False) == (
+            False,
+            infra.jsonrpc.ErrorCode.CODE_ID_NOT_FOUND,
+        )
+
+        add_new_code(primary)
+
+        with open("networkcert.pem", mode="rb") as file:
+            net_cert = list(file.read())
+
+        new_nodes = set()
+        # add nodes using the same code id that failed earlier
+        for i in range(4, 9):
+            LOG.debug(f"Adding node {i} using new code")
+            res, new_node, new_node_id = network.create_and_add_node(
+                args.patched_file_name, args, i, True
+            )
+            assert res
+            new_node.join_network()
+            new_nodes.add(new_node)
+
+        network.wait_for_node_commit_sync()
+
+        for node in new_nodes:
```

Comment:
I think this is not the a guaranteed way to find the primary. Instead, we should wait until the old nodes have been stopped and some time has elapsed and then run [network.find_leader()](https://github.com/microsoft/CCF/blob/7c09e1ec4e4a3461afd120fbd6f37dc3556febf3/tests/infra/ccf.py#L293). I believe that what's happening now is that node 13's join request is forwarded to the leader anyway if `new_primary` is not the leader (and that works just fine).